### ISSUE TITLE
Adding custom properties to model.json upon save?

### ISSUE THREAD

**Issue by shiffman** [tensorflow/tfjs#762]
Hello! I'm writing on behalf of the [ml5](https://ml5js.org/) project. In a meeting today with @cvalenzuela and other contributors we discussed some questions around the saving and loading features for ml5. This is discussed in more detail at https://github.com/ml5js/ml5-library/issues/174.

We are wrapping the tf.js `save()` method in our [ImageClassifier](https://ml5js.org/docs/ImageClassifier.html) class for use with the [FeatureExtractor](https://ml5js.org/docs/FeatureExtractor) example. `model.json` and `model.weights.bin` work "out of the box" with ml5, however our examples include additional functionality. For example, we allow the end user to specify a string label for a class rather than a numeric index. To be able to retain this information when saving, we would like to be able to customize `model.json` with ml5 specific properties.

What do you think about adding an optional argument to `save()` that allows for customization, e.g.

```javascript
const ml5prop = { 
  labels: ['cat', 'dog']
};
model.save(destination, ml5prop);
```

This could be good for other use cases where tf.js users have other meta-data they would like to include.

Let us know your thoughts. We are happy to work on a pull request for this. Or perhaps you have a better idea for how we should be handling saving/loading on our end?

@cvalenzuela let me know if I missed anything or am not describing this properly!

**Comment by caisq** [comment]
I think this is a reasonable feature request.

**Comment by davidsoergel** [comment]
Are the types of metadata that you want to include fairly structured, slowly changing, & generally useful?  If so I'd prefer to establish standards for a specific set of fields.  The label index to string mapping is certainly something that should be standardized.

I fear that if we allow arbitrary extensions to model.json, the result will be a profusion of mutually-incompatible file formats.  For that matter I think Python Keras will currently error out on nonstandard fields.

So to me this is an ecosystem-wide question of getting the file formats right, not something we should just do as a local convenience.  That said, if there is consensus that model.json can have a 'custom_metadata' field that loaders can know to ignore, that would be at least non-breaking (but would still not provide the benefit of standardization).

Another workaround you could do right away would be to wrap model.save() in another function, e.g. saveModelWithMetadata(model, ml5prop) that writes a separate metadata.json file.  That might require updates to IOHandlers (basically to allow writing and reading additional files, analogous to the assets/ directory in TF SavedModel).

**Comment by cvalenzuela** [comment]
Perhaps the idea of `saveModelWithMetadata()` might work better for compatibility formats across the ecosystem. 

But building on top of that, maybe something like `model.saveWeights()`  in conjunction with `model.getSpecs()` could also be implemented? `model.saveWeights()` can be used just to get the weights file and `model.getSpecs()` can return a json with specs that we can then internally modify and then trigger to download. I have had issues when downloading multiple files with `model.save()` in Chrome (https://bugs.chromium.org/p/chromium/issues/detail?id=822542) so having options to download the files individually could potentially help?

**Comment by nsthorat** [comment]
@davidsoergel right now the model.json is already a "custom" format, given that it marries the weights manifest file and the actual model JSON. Adding another valid JSON field seems okay because it's specific to this environment anyways.

**Comment by davidsoergel** [comment]
@nsthorat as you know, I think we should not have diverged from keras.json in the first place.  But this would go much further.  1) model.json was *one* new format.  The proposal here would create *arbitrarily many* new formats because each user / library developer would store metadata in an idiosyncratic form in the new custom field.  2) Many useful kinds of metadata are not specific to any particular environment.  The given example, mapping class IDs to string labels, is important in every environment--so it would be sad if tool A writes it in a form that tool B can't read.

So it seems to me that this approach encourages balkanization, where we should be encouraging interoperability.  Let's channel this motivation to store metadata (with which I wholeheartedly agree!) into a standard representation.  In particular, I think it would be great to work with the tf.Metadata team (https://github.com/tensorflow/metadata, @mzinkevi) to make sure they publish enough docs / specs / JS bindings that we can integrate it here.

**Comment by shiffman** [comment]
@cvalenzuela and I were discussing this a bit more on Friday. In case this is helpful, we would be very happy to do our own file saving and loading in ml5 and not have any changes to `model.save()` in tf.js. The issue right now is the only way for us to get the data that ends up in `model.json` is by saving a file. If there were a method like `model.getJSON()` that returned what ends up in the saved file, we could make our own modifications and save our own file. Thanks so much for this discussion!

**Comment by davidsoergel** [comment]
ha you just beat me to it :)

In particular: you can use withSaveHandler (https://github.com/tensorflow/tfjs-core/blob/master/src/io/passthrough.ts#L96), where you can provide a callback that receives the to-be-saved objects as input.

**Comment by shiffman** [comment]
@nsthorat @davidsoergel oh, I hadn't noticed that when looking over the docs! Thank you! @cvalenzuela, will that work for our needs? If so, we can close this issue (unless it's a relevant discussion for other aspects of the project too).

**Comment by davidsoergel** [comment]
Looks like withSaveHandler() is not mentioned in the generated API docs.  Let's leave this issue open until that is fixed, at least.

**Comment by cvalenzuela** [comment]
nice! `withSaveHandler()` seems to be what we need. I'll try that instead of `model.save()`

**Comment by cvalenzuela** [comment]
I'm using `withSaveHandler()` in the following way:

```javascript
model.save(tf.io.withSaveHandler((obj2save) => { ... }))
```

and the callback value, `obj2save`, being returned is: 

```javascript
{
    modelTopology: {…}, 
    weightData: ArrayBuffer(5018800), 
    weightSpecs: Array(3)
}
```

What's the best way to save the `weightData` into a `.bin` file?  What's the difference between `weightSpecs` generated with `withSaveHandler` and `weightsManifest` that you get with `.save('downloads://')`?

**Comment by nsthorat** [comment]
weightSpecs are a subset of weightsManifest that don't contain the paths. weightsSpecs are of type `WeightsManifestEntry`, here is the corresponding entry in the weights manifest: https://github.com/tensorflow/tfjs-core/blob/master/src/io/types.ts#L59

You can just kick off a download using a Blob, createObjectURL and forcing a link (google how to trigger file download from JS).

**Comment by rthadur** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by jackrobertscott** [comment]
Here is a work around:

```ts
await model.save(
  tf.io.http(`https://myserver.com/UploadModel`, {
    fetchFunc: (path: string, init: RequestInit) => {
      if (init.body instanceof FormData)
        init.body.set('example', 'Hello world!')
      return window.fetch(path, init)
    },
  })
)
```

The above code sets the `FormData` property of `example` to `Hello world!` which can be extracted on the server.

**Comment by lina128** [comment]
Hi @jackrobertscott ,
If you want to store metadata during conversion, see the --metadata option from converter (https://github.com/tensorflow/tfjs/tree/master/tfjs-converter): 
```
--metadata | Comma separated list of metadata json file paths, indexed by name. Prefer absolute path. Example: 'metadata1:/metadata1.json,metadata2:/metadata2.json'.
```

If you want to store metadata to model in memory and save it, you can store objects to userDefinedMetadata map: https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/io/types.ts#L289

Does it fulfill your needs?

**Comment by gaberogan** [comment]
@rthadur Can we reopen this? It would be useful to have `withSaveHandler` in the docs. Thanks.

**Comment by gaikwadrahul8** [comment]
Hi, @shiffman 

Apologize for the delayed response and I see we have updated official documentation w.r.t save method [save](https://js.tensorflow.org/api/latest/#tf.GraphModel.save) and also this PR https://github.com/tensorflow/tfjs/pull/4187 got merged.

@gaberogan please refer updated [official documentation](https://js.tensorflow.org/api/latest/#tf.GraphModel.save) 

If we have taken care of your issue, could you please feel free to close this issue ? Thank you!
